format/nes/nes.go

```diff
@@ -0,0 +1,339 @@
+	5:  {Description: "V.R. Technology VT01 with red/cyan STN palette"},
+	6:  {Description: "V.R. Technology VT02"},
+	7:  {Description: "V.R. Technology VT03"},
+	8:  {Description: "V.R. Technology VT09"},
+	9:  {Description: "V.R. Technology VT32"},
+	10: {Description: "V.R. Technology VT369"},
+	11: {Description: "UMC UM6578"},
+	12: {Description: "Famicom Network System"},
+	13: {Description: "reserved"},
+	14: {Description: "reserved"},
+	15: {Description: "reserved"},
+}
+
+func decodeFileHeader(d *decode.D, nc *nesContext) {
+	// byte 0-3
+	d.FieldRawLen("identifier", 4*8, d.AssertBitBuf([]byte{0x4e, 0x45, 0x53, 0x1a})) // NES<EOF>
+
+	// byte 4
+	prgLSB := d.FieldU8("prg_rom_size_lsb")
+
+	// byte 5
+	chrLSB := d.FieldU8("chr_rom_size_lsb")
+
+	// byte 6
+	d.FieldU1("nametable_layout", bitDesc("Vertical arrangement", "Horizontal arrangement"))
+	d.FieldBool("battery", boolDesc("Not present", "Present"))
+	trainer := d.FieldBool("trainer", boolDesc("Not present", "Present"))
+	d.FieldBool("alternative_nametables")
+	mapperLSB := d.FieldU4("mapper_lsb")
+
+	// byte 7 (half)
+	consoleType := d.FieldU2("console_type", consoleTypeMapper)
+	nes20 := d.FieldU2("nes20_identifier", bitDesc("iNES", "NES 2.0"))
+
+	if nes20 == 2 { // NES 2.0
+		// byte 7 second half and byte 8
+		mapperMSB := d.FieldU8("mapper_msb")
+		d.FieldValueUint("mapper_full", mapperLSB+(mapperMSB<<4))
+		d.FieldU4("submapper")
+
+		// byte 9
+		prgMSB := d.FieldU4("prg_rom_size_msb")
+		chrMSB := d.FieldU4("chr_rom_size_msb")
+		prgROMSize := romSize(prgLSB, prgMSB, 16)
+		d.FieldValueUint("prg_rom_size_full", prgROMSize)
+		chrROMSize := romSize(chrLSB, chrMSB, 8)
+		d.FieldValueUint("chr_rom_size_full", chrROMSize)
+
+		shiftSizeMapper := getShiftSizeMapper()
+		// byte 10
+		d.FieldU4("prg_ram_size", shiftSizeMapper)
+		d.FieldU4("prg_nvram_size", shiftSizeMapper)
+
+		// byte 11
+		d.FieldU4("chr_ram_size", shiftSizeMapper)
+		d.FieldU4("chr_nvram_size", shiftSizeMapper)
+
+		// byte 12
+		d.FieldU2("cpu_ppu_timing_mode", timingModeMapper)
+		d.SeekRel(6)
+
+		// byte 13
+		switch consoleType {
+		case 1:
+			d.FieldU4("vs_ppu_type", vsPPUMapper)
+			d.FieldU4("vs_hardware_type", vsHardwareMapper)
+		case 3:
+			d.FieldU4("ext_console_type", extConsoleMapper)
+			d.SeekRel(4)
+		default:
+			d.SeekRel(8)
+		}
+
+		// byte 14
+		d.FieldU2("misc_roms")
+		d.SeekRel(6)
+
+		// byte 15
+		d.FieldU6("default_exp_device")
+		d.SeekRel(2)
+
+		nc.prgROMSize = prgROMSize
+		nc.chrROMSize = chrROMSize
+	} else { // iNES
+		// byte 7 (second half)
+		mapperMSB := d.FieldU4("mapper_msb")
+		d.FieldValueUint("mapper_full", mapperLSB+(mapperMSB<<4))
+
+		prgROMSize := romSize(prgLSB, 0, 16)
+		d.FieldValueUint("prg_rom_size_full", prgROMSize)
+		chrROMSize := romSize(chrLSB, 0, 8)
+		d.FieldValueUint("chr_rom_size_full", chrROMSize)
+
+		// byte 8
+		d.FieldU8("prg_ram_size", ramSizer(8))
+
+		// byte 9 - 15
+		d.FieldStr("crap", 7, encoding.Nop)
```

Comment:
Haven't seen `encoding.Nop` before. I wonder if it should be used in some other places in fq. I've thought about having maybe `d.FieldASCII` etc, i guss that would be nop?

Reply:
I was looking for an ASCII, but found this first, and since it is just padding that people sometimes puts text in I figured it was good enough.

Reply:
Aha i see, makes sense